k8s scan compare results

```
FTL failed to resolve policies error="failed to compile bundle: failed to compile //local.cnspec.io/run/local-execution/queries/mondoo-kubernetes-best-practices-ingress-cert-expiration: failed to compile query 'k8s.ingress.tls.all(\n  certificates.all(\n    expiresIn.days > 15\n  )\n)\n': cannot find resource that is called by '>' of type core.certificate\nfailed to compile //local.cnspec.io/run/local-execution/queries/mondoo-kubernetes-best-practices-ingress-cert-expiration: failed to compile query 'k8s.ingress.tls.all(\n  certificates.all(\n    expiresIn.days > 15\n  )\n)\n': cannot find resource that is called by '>' of type core.certificate\n"
exit status 1
```

I couldn't really pin this down to a query. I get a panic, because the cluster doesn't have certificates:
```
cnquery run k8s --discover ingresses -c "k8s.ingress.tls{ certificates{ * }}" 
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0xc4ecbf]

goroutine 1 [running]:
go.mondoo.com/cnquery/mqlc.publicFieldsInfo(0xc000757a70, 0x0)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/builtin.go:192 +0x3f
go.mondoo.com/cnquery/mqlc.availableGlobFields(0xc000757a70, {0xc000774de1?, 0x0?}, 0x0)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/builtin.go:257 +0xc7
go.mondoo.com/cnquery/mqlc.(*compiler).compileBlock(0xc000757a70, {0xc00007ea28?, 0x1?, 0x1?}, {0xc000774de0, 0x12}, 0x0?)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:292 +0x15c
go.mondoo.com/cnquery/mqlc.(*compiler).compileOperand(0xc000757a70, 0xc00075d3b0)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:1580 +0xdcc
go.mondoo.com/cnquery/mqlc.(*compiler).compileExpression(...)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:1596
go.mondoo.com/cnquery/mqlc.(*compiler).compileAndAddExpression(0xc000757a70, 0xcc67bdf3862d5bdf?)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:1600 +0x2b
go.mondoo.com/cnquery/mqlc.(*compiler).compileExpressions(0xc000757a70, {0xc00007ea30, 0x1, 0x1?})
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:1676 +0x2ad
go.mondoo.com/cnquery/mqlc.(*compiler).blockOnResource(0xc0007579e0, {0xc00007ea30, 0x1, 0x1}, {0xc00072fcc1, 0xf}, 0x100000002)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:590 +0x333
go.mondoo.com/cnquery/mqlc.(*compiler).blockExpressions(0xc0007579e0, {0xc00007ea30, 0x1, 0x1}, {0xc00072fcc0?, 0xc000856e50?}, 0x100000002)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:662 +0x30a
go.mondoo.com/cnquery/mqlc.(*compiler).compileBlock(0xc0007579e0, {0xc00007ea30?, 0x1?, 0x1?}, {0xc00072fcc0, 0x10}, 0x2?)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:310 +0x197
go.mondoo.com/cnquery/mqlc.(*compiler).compileOperand(0xc0007579e0, 0xc00075d310)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:1580 +0xdcc
go.mondoo.com/cnquery/mqlc.(*compiler).compileExpression(...)
	/home/christian/workspace/mondoo/github.com/cnquery/mqlc/mqlc.go:1596
```

One level higher als is fine:
```
cnquery run k8s --discover ingresses -c "k8s.ingress.tls{ * }" 
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
k8s.ingress.tls: []
k8s.ingress.tls: []
```

It's getting wired. In a cluster with an ingress including a tls section, I get a panic:
```
cnquery run k8s --discover ingresses -c "k8s.ingress.tls" 
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0xe26ea7]

goroutine 73 [running]:
go.mondoo.com/cnquery/providers.(*Runtime).lookupResourceProvider(0xc0005269c0, {0xc000659c60, 0xc})
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:537 +0x187
go.mondoo.com/cnquery/providers.(*Runtime).CreateResource(0xc0005269c0, {0xc000659c60, 0xc}, 0xc000c01860)
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:228 +0x3f
go.mondoo.com/cnquery/providers.(*providerCallbacks).GetData(0x128d600?, 0x412600?)
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:402 +0x45
go.mondoo.com/cnquery/providers-sdk/v1/plugin.(*GRPCProviderCallbackServer).GetData(0x1363ac0?, {0xc0001fcd90?, 0x49ea26?}, 0x0?)
	/home/christian/workspace/mondoo/github.com/cnquery/providers-sdk/v1/plugin/grpc.go:154 +0x1e
go.mondoo.com/cnquery/providers-sdk/v1/plugin._ProviderCallback_GetData_Handler({0x128d600?, 0xc000a32600}, {0x1607830, 0xc000c01470}, 0xc0001fccb0, 0x0)
	/home/christian/workspace/mondoo/github.com/cnquery/providers-sdk/v1/plugin/plugin_grpc.pb.go:427 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0006b0000, {0x160dc60, 0xc0006a2680}, 0xc000391680, 0xc000968de0, 0x1fd6250, 0x0)
	/home/christian/go/pkg/mod/google.golang.org/grpc@v1.58.1/server.go:1376 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc0006b0000, {0x160dc60, 0xc0006a2680}, 0xc000391680, 0x0)
	/home/christian/go/pkg/mod/google.golang.org/grpc@v1.58.1/server.go:1753 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	/home/christian/go/pkg/mod/google.golang.org/grpc@v1.58.1/server.go:998 +0x8d
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 101
	/home/christian/go/pkg/mod/google.golang.org/grpc@v1.58.1/server.go:996 +0x165

 ~/workspace/mondoo/github.com/cnquery │ main *8 ?4  k get ingress 
NAME                CLASS   HOSTS              ADDRESS   PORTS     AGE
hello-app-ingress   nginx   demo.cz-test.com             80, 443   3m19s
```

--- a/providers/k8s/resources/ingress.go
+++ b/providers/k8s/resources/ingress.go
@@ -20,10 +20,9 @@ import (
 )
 
 type mqlK8sIngressInternal struct {
-	lock   sync.Mutex
-	obj    *networkingv1.Ingress
-	objId  string
-	mqlK8s *mqlK8s
+	lock  sync.Mutex
+	obj   *networkingv1.Ingress
+	objId string
 }
 
 func (k *mqlK8s) ingresses() ([]interface{}, error) {
@@ -63,15 +62,20 @@ func (k *mqlK8s) ingresses() ([]interface{}, error) {
 		}
 		r.(*mqlK8sIngress).obj = ingress
 		r.(*mqlK8sIngress).objId = objId
-		r.(*mqlK8sIngress).mqlK8s = k
 		return r, nil
 	})
 }
 
 func (k *mqlK8sIngress) tls() ([]interface{}, error) {
+	o, err := CreateResource(k.MqlRuntime, "k8s", map[string]*llx.RawData{})
+	if err != nil {
+		return nil, err
+	}
+	k8s := o.(*mqlK8s)
+
 	ingress := k.obj
 	objId := k.objId
-	tls, err := getTLS(ingress, objId, k.MqlRuntime, k.mqlK8s.GetSecrets)
+	tls, err := getTLS(ingress, objId, k.MqlRuntime, k8s.GetSecrets)
 	if err != nil {
 		return nil, err
 	}

--- a/providers/k8s/resources/ingress.go
+++ b/providers/k8s/resources/ingress.go
@@ -20,8 +20,10 @@ import (
 )
 
 type mqlK8sIngressInternal struct {
-	lock sync.Mutex
-	obj  *networkingv1.Ingress
+	lock   sync.Mutex
+	obj    *networkingv1.Ingress
+	objId  string
+	mqlK8s *mqlK8s
 }
 
 func (k *mqlK8s) ingresses() ([]interface{}, error) {
@@ -45,11 +47,6 @@ func (k *mqlK8s) ingresses() ([]interface{}, error) {
 			return nil, err
 		}
 
-		tls, err := getTLS(ingress, objId, k.MqlRuntime, k.GetSecrets)
-		if err != nil {
-			return nil, err
-		}
-
 		r, err := CreateResource(k.MqlRuntime, "k8s.ingress", map[string]*llx.RawData{
 			"id":              llx.StringData(objId),
 			"uid":             llx.StringData(string(obj.GetUID())),
@@ -60,14 +57,26 @@ func (k *mqlK8s) ingresses() ([]interface{}, error) {
 			"created":         llx.TimeData(ts.Time),
 			"manifest":        llx.DictData(manifest),
 			"rules":           llx.ArrayData(rules, types.Resource("k8s.ingressrule")),
-			"tls":             llx.ArrayData(tls, types.Resource("k8s.ingresstls")),
 		})
 		if err != nil {
 			return nil, err
 		}
 		r.(*mqlK8sIngress).obj = ingress
+		r.(*mqlK8sIngress).objId = objId
+		r.(*mqlK8sIngress).mqlK8s = k
 		return r, nil
 	})
+}
+
+func (k *mqlK8sIngress) tls() ([]interface{}, error) {
+	ingress := k.obj
+	objId := k.objId
+	tls, err := getTLS(ingress, objId, k.MqlRuntime, k.mqlK8s.GetSecrets)
+	if err != nil {
+		return nil, err
+	}
+
+	return tls, nil
 }
 
 func (k *mqlK8sIngress) id() (string, error) {
@@ -296,7 +305,7 @@ func getTLS(ingress *networkingv1.Ingress, objId string, runtime *plugin.Runtime
 			ingressTls, err := CreateResource(runtime, "k8s.ingresstls", map[string]*llx.RawData{
 				"id":           llx.StringData(fmt.Sprintf("%s-tls%d", objId, i)),
 				"hosts":        llx.ArrayData(convert.SliceAnyToInterface(tls.Hosts), types.String),
-				"certificates": llx.ArrayData(secret.Certificates.Data, types.Resource("core.certificate")),
+				"certificates": llx.ArrayData(secret.Certificates.Data, types.Resource("network.certificate")),
 			})
 			if err != nil {
 				return nil, fmt.Errorf("error creating k8s.ingresstls: %s", err)

--- a/providers/k8s/resources/ingress.go
+++ b/providers/k8s/resources/ingress.go
@@ -73,9 +73,7 @@ func (k *mqlK8sIngress) tls() ([]interface{}, error) {
 	}
 	k8s := o.(*mqlK8s)
 
-	ingress := k.obj
-	objId := k.objId
-	tls, err := getTLS(ingress, objId, k.MqlRuntime, k8s.GetSecrets)
+	tls, err := getTLS(k.obj, k.objId, k.MqlRuntime, k8s.GetSecrets)
 	if err != nil {
 		return nil, err
 	}

--- a/providers/k8s/resources/k8s.lr
+++ b/providers/k8s/resources/k8s.lr
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: BUSL-1.1
 
 import "../../os/resources/os.lr"
+import "../../network/resources/network.lr"
 
 option provider = "go.mondoo.com/cnquery/providers/k8s"
 option go_package = "go.mondoo.com/cnquery/providers/k8s/resources"
@@ -465,7 +466,7 @@ private k8s.secret @defaults("namespace name created") {
   // Secret type
   type string
   // Secret certificates
-  certificates() []certificate
+  certificates() []network.certificate
 }
 
 // Kubernetes ConfigMap
@@ -583,7 +584,7 @@ private k8s.ingresstls {
   // List of hosts associated with TLS certificate
   hosts []string
   // Certificates data from TLS Secret
-  certificates []core.certificate
+  certificates []network.certificate
 }
 
 // Kubernetes Ingress
@@ -611,7 +612,7 @@ private k8s.ingress @defaults("namespace name created") {
   // Ingress rules
   rules []k8s.ingressrule
   // Ingress TLS data
-  tls []k8s.ingresstls
+  tls() []k8s.ingresstls
 }
 
 // Kubernetes Service Account

--- a/providers/k8s/resources/k8s.lr.go
+++ b/providers/k8s/resources/k8s.lr.go
@@ -956,7 +956,7 @@ var getDataFields = map[string]func(r plugin.Resource) *plugin.DataRes{
 		return (r.(*mqlK8sIngresstls).GetHosts()).ToDataRes(types.Array(types.String))
 	},
 	"k8s.ingresstls.certificates": func(r plugin.Resource) *plugin.DataRes {
-		return (r.(*mqlK8sIngresstls).GetCertificates()).ToDataRes(types.Array(types.Resource("core.certificate")))
+		return (r.(*mqlK8sIngresstls).GetCertificates()).ToDataRes(types.Array(types.Resource("certificate")))
 	},
 	"k8s.ingress.id": func(r plugin.Resource) *plugin.DataRes {
 		return (r.(*mqlK8sIngress).GetId()).ToDataRes(types.String)
@@ -5766,7 +5766,19 @@ func (c *mqlK8sIngress) GetRules() *plugin.TValue[[]interface{}] {
 }
 
 func (c *mqlK8sIngress) GetTls() *plugin.TValue[[]interface{}] {
-	return &c.Tls
+	return plugin.GetOrCompute[[]interface{}](&c.Tls, func() ([]interface{}, error) {
+		if c.MqlRuntime.HasRecording {
+			d, err := c.MqlRuntime.FieldResourceFromRecording("k8s.ingress", c.__id, "tls")
+			if err != nil {
+				return nil, err
+			}
+			if d != nil {
+				return d.Value.([]interface{}), nil
+			}
+		}
+
+		return c.tls()
+	})
 }
 
 // mqlK8sServiceaccount for the k8s.serviceaccount resource

